src/context-popup/index.tsx

```diff
@@ -0,0 +1,82 @@
+import { create, tsx } from '@dojo/framework/core/vdom';
+import focus from '@dojo/framework/core/middleware/focus';
+import { RenderResult } from '@dojo/framework/core/interfaces';
+import Popup from '../popup';
+import { createICacheMiddleware } from '@dojo/framework/core/middleware/icache';
+
+import * as css from '../theme/default/context-popup.m.css';
+
+export interface ContextPopupProperties {
+	onClose?(): void;
+	onOpen?(): void;
+}
+
+export interface ContextPopupChildren {
+	contentWithContext: () => RenderResult;
+	popupContent: (close: () => void, shouldFocus: () => boolean) => RenderResult;
+}
+
+interface ContextIcache {
+	x: number;
+	y: number;
+	open: boolean;
+}
+
+const icache = createICacheMiddleware<ContextIcache>();
+
+const factory = create({ icache, focus })
+	.properties<ContextPopupProperties>()
+	.children<ContextPopupChildren>();
+
+const CursorWidth = 2;
+const CursorHeight = 4;
+
+export const ContextPopup = factory(function({
+	properties,
+	children,
+	middleware: { icache, focus }
+}) {
+	const { onOpen, onClose } = properties();
+	const x = icache.getOrSet('x', 0);
+	const y = icache.getOrSet('y', 0);
+
+	const { contentWithContext, popupContent } = children()[0];
+	const close = () => {
+		icache.set('open', false);
+		onClose && onClose();
+	};
+
+	return (
+		<virtual>
+			<div
+				classes={css.contentWithContext}
+				key="contentWithContext"
+				oncontextmenu={(event: MouseEvent) => {
+					event.preventDefault();
+					focus.focus();
+					icache.set('x', event.pageX - CursorWidth);
+					icache.set('y', event.pageY - CursorHeight);
+					icache.set('open', true);
+					onOpen && onOpen();
+				}}
+			>
+				{contentWithContext()}
+			</div>
+			<Popup
+				key="popup"
+				x={x}
+				yTop={y}
+				yBottom={document.documentElement.scrollTop + document.documentElement.clientHeight}
+				onClose={close}
+				position="below"
+				open={icache.get('open')}
+			>
+				{{
+					content: () => <div>{popupContent(close, focus.shouldFocus)}</div>
```

Comment:
Why does this need to pass `should focus`?  This is done manually for the select and other such popup's, is it required here due to the nature of the trigger / right click?

Reply:
The idea is that when you use this popup you can just pass `shouldFocus` to the `focus` property of whatever element in your popup should receive the initial focus. `Select` is already rendering the `Menu` element. On my context-menu branch I've applied this to the `Menu` rendered in the context popup.

Reply:
does the focus not need to be controlled by whatever creates both the trigger and the popup content though so it can chose to pass the focus between them? I guess it's probably better this way than manually doing it for every context menu. Just wanted to clarify